### PR TITLE
Addition of Source Terms for NEMO Axisymmetric Flows

### PR DESCRIPTION
**Pull request by jtneedels**
## Proposed Changes
 

- Additional source terms are implemented in the residual computation for NEMO planar axisymmetric flows. This allows inviscid and viscous axisymmetric simulations. 

## Related Work

- Built upon framework in PR #1106 by FlorianDM, which introduced viscous axisymmetric source terms for non-reacting flows.

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags, or simply --warnlevel=2 when using meson).
- [x] My contribution is commented and consistent with SU2 style.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

### CODE REVIEW

File: SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp [hunk 1]
```diff
@@ -296,7 +296,24 @@ CNumerics::ResidualType<> CSource_NEMO::ComputeVibRelaxation(const CConfig *conf
 CNumerics::ResidualType<> CSource_NEMO::ComputeAxisymmetric(const CConfig *config) {
 
   unsigned short iDim, iSpecies, iVar;
-  su2double rho, rhov, vel2, H, yinv;
+  su2double rho, rhov, vel2, H, yinv, T, Tve, qy_ve, Ru, RuSI;
+  su2double *Ds, **GV, ktr, kve;
+
+  Ds = Diffusion_Coeff_i;
```
Comment (WallyMaier) [inline SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp:302]: Maybe add a comment here.  Not necessary but I think looks better.   i.e.  "Renaming for conveniences"

File: SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp [hunk 2]
```diff
@@ -324,6 +342,45 @@ CNumerics::ResidualType<> CSource_NEMO::ComputeAxisymmetric(const CConfig *confi
   residual[nSpecies+2] = yinv*rhov*H*Volume;
   residual[nSpecies+3] = yinv*rhov*U_i[nSpecies+nDim+1]/rho*Volume;
 
+     if (viscous) {
```
Comment (WallyMaier) [inline SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp:345]: This spacing looks off.  Extra spaces?

File: SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp [hunk 3]
```diff
@@ -324,6 +342,45 @@ CNumerics::ResidualType<> CSource_NEMO::ComputeAxisymmetric(const CConfig *confi
...
+        for (iSpecies=0; iSpecies<nSpecies; iSpecies++)
+          Mass += V_i[iSpecies]*Ms[iSpecies];
+        
+        su2double heat_capacity_cp_i   = V_i[RHOCVTR_INDEX]+Ru/Mass;
+        su2double total_viscosity_i    = Laminar_Viscosity_i + Eddy_Viscosity_i;
+        su2double total_conductivity_i = ktr + kve + heat_capacity_cp_i*Eddy_Viscosity_i/Prandtl_Turb;      
+        su2double u = V_i[VEL_INDEX];
+        su2double v = V_i[VEL_INDEX+1];
+
+        qy_ve = kve*GV[TVE_INDEX][1];
```
Comment (WallyMaier) [inline SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp:362]: you could do something similar here: "su2double qy_ve"

Again, this just aesthetic

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp [hunk 4]
```diff
@@ -1150,49 +1153,103 @@ void CNEMOEulerSolver::Source_Residual(CGeometry *geometry, CSolver **solver_con
...
+        /*--- Set control volume ---*/
+        numerics->SetVolume(geometry->nodes->GetVolume(iPoint));
+
+        /*--- Set y coordinate ---*/
+        numerics->SetCoord(geometry->nodes->GetCoord(iPoint), geometry->nodes->GetCoord(iPoint));
+
+        /*--- Set primitive variables for viscous terms and/or generalised source ---*/
+        numerics->SetPrimitive(nodes->GetPrimitive(iPoint), nodes->GetPrimitive(iPoint));
+
+        if (viscous) {
```
Comment (WallyMaier) [inline SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:1199]: maybe add a comment here

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp [hunk 5]
```diff
@@ -1150,49 +1153,103 @@ void CNEMOEulerSolver::Source_Residual(CGeometry *geometry, CSolver **solver_con
...
+          numerics->SetEve(nodes->GetEve(iPoint), nodes->GetEve(iPoint));
+
+          /*--- Set turbulence kinetic energy ---*/
+          if (rans){
+            CVariable* turbNodes = solver_container[TURB_SOL]->GetNodes();
+            numerics->SetTurbKineticEnergy(turbNodes->GetSolution(iPoint,0), turbNodes->GetSolution(iPoint,0));
+          }
+        }
+
+          auto residual = numerics->ComputeAxisymmetric(config);
```
Comment (WallyMaier) [inline SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:1232]: Something is off with your spacings here

File: SU2_CFD/src/variables/CNEMOEulerVariable.cpp [hunk 6]
```diff
@@ -78,6 +78,14 @@ CNEMOEulerVariable::CNEMOEulerVariable(su2double val_pressure,
 
   Res_TruncError.resize(nPoint,nVar) = su2double(0.0);
 
+    /*--- Size Grad_AuxVar for axiysmmetric ---*/
+
```
Comment (WallyMaier) [inline SU2_CFD/src/variables/CNEMOEulerVariable.cpp:82]: spacings

**Review by WallyMaier — CHANGES_REQUESTED** [review]
@jtneedels thanks for deriving these source terms.  They will be very helpful! 

Please find my comments above

File: SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp (hunk 1 above)
Comment (jtneedels) [reply SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp:302]: Fixed.

Comment (jtneedels) [reply SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp:303]: Okay, addressed this.

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp (hunk 4 above)
Comment (jtneedels) [reply SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:1200]: Is this what you mean by a comment here?

File: SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp (hunk 1 above)
Comment (WallyMaier) [reply SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp:303]: Thanks!

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp (hunk 4 above)
Comment (WallyMaier) [reply SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:1200]: Yup!

**Review by WallyMaier — APPROVED** [review]
This looks good to me!  I think you should wait for another approval before pushing.

**Comment by pcarruscag** [comment]
@jtneedels the testcases currently failing are my fault, they will be fixed in a minute.

File: SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp [hunk 7]
```diff
@@ -324,6 +343,44 @@ CNumerics::ResidualType<> CSource_NEMO::ComputeAxisymmetric(const CConfig *confi
   residual[nSpecies+2] = yinv*rhov*H*Volume;
   residual[nSpecies+3] = yinv*rhov*U_i[nSpecies+nDim+1]/rho*Volume;
 
+  if (viscous) {
+
```
Comment (CatarinaGarbacz) [inline SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp:347]: extra line here

File: SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp [hunk 8]
```diff
@@ -324,6 +343,44 @@ CNumerics::ResidualType<> CSource_NEMO::ComputeAxisymmetric(const CConfig *confi
...
+    su2double total_conductivity_i = ktr + kve + heat_capacity_cp_i*Eddy_Viscosity_i/Prandtl_Turb;      
+    su2double u                    = V_i[VEL_INDEX];
+    su2double v                    = V_i[VEL_INDEX+1];
+    su2double qy_ve                = kve*GV[TVE_INDEX][1];
+
+    /*--- Enthalpy and vib-el energy transport due to y-direction diffusion---*/
+    for (iSpecies = 0; iSpecies < nHeavy; iSpecies++) { 
+      sumJhs_y += (rho*Ds[iSpecies]*GV[RHOS_INDEX+iSpecies][1] - V_i[RHOS_INDEX+iSpecies]*Vector[1]) * hs[iSpecies];
+      sumJeve_y += (rho*Ds[iSpecies]*GV[RHOS_INDEX+iSpecies][1] - V_i[RHOS_INDEX+iSpecies]*Vector[1]) * eve_i[iSpecies];
+    }
```
Comment (CatarinaGarbacz) [inline SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp:368]: you can just ident this 2 lines for aesthetics (the += part)

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp [hunk 9]
```diff
@@ -1150,25 +1153,103 @@ void CNEMOEulerSolver::Source_Residual(CGeometry *geometry, CSolver **solver_con
...
+        /*--- Apply the update to the linear system ---*/
+        if (!err) {
+          LinSysRes.AddBlock(iPoint, residual);
+          if (implicit)
+            Jacobian.AddBlock(iPoint, iPoint, Jacobian_i);
+        }else
+          eAxi_local++;
+      }
+
     }
```
Comment (CatarinaGarbacz) [inline SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:1253]: extra line between between the brackets

**Review by CatarinaGarbacz — APPROVED** [review]
Hi @jtneedels . good work. I just made some minor comments, I didn't notice anything "big", so for me it's a GO

File: SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp (hunk 8 above)
Comment (pcarruscag) [reply SU2_CFD/src/numerics/NEMO/NEMO_sources.cpp:368]: I would not spend too much time with that kind of alignment, eventually we will run clang-format on the entire code... The only thing you can do to make code look good is to write small functions.

File: SU2_CFD/src/solvers/CNEMOEulerSolver.cpp (hunk 9 above)
Comment (jtneedels) [reply SU2_CFD/src/solvers/CNEMOEulerSolver.cpp:1252]: Thanks for the review Catarina, I think I've addressed this and the other comments on formatting.

**Review by fmpmorgado — APPROVED** [review]

**Pull request merged**